src/mrinufft/operators/off_resonance.py

```diff
@@ -0,0 +1,330 @@
+
+        # get bin centers
+        zc = [e[1:] - (e[1] - e[0]) / 2 for e in ze]
+
+        # complexify
+        zk = _outer_sum(1j * zc[0], zc[1])  # [K1 K2]
+        zk = zk.T
+        hk = hk.T
+    else:
+        z = fieldmap[mask].ravel()
+        hk, ze = xp.histogram(z.imag, bins=n_bins[0])
+
+        # get bin centers
+        zc = ze[1:] - (ze[1] - ze[0]) / 2
+
+        # complexify
+        zk = 0 + 1j * zc  # [K 1]
+
+    # flatten histogram values and centers
+    hk = hk.ravel()
+    zk = zk.ravel()
+
+    # generate time for each segment
+    tl = xp.linspace(
+        readout_time.min(), readout_time.max(), n_time_segments, dtype=xp.float32
+    )  # time seg centers in [s]
+
+    # prepare for basis calculation
+    ch = xp.exp(-tl[:, None, ...] @ zk[None, ...])
+    w = xp.diag(hk**0.5)
+    p = xp.linalg.pinv(w @ ch.T) @ w
+
+    # actual temporal basis calculation
+    B = p @ xp.exp(-zk[:, None, ...] * readout_time[None, ...])
+    B = B.astype(xp.complex64)
+
+    # get spatial coeffs
+    C = xp.exp(-tl * fieldmap[..., None])
+    C = C[None, ...].swapaxes(0, -1)[
+        ..., 0
+    ]  # (..., n_time_segments) -> (n_time_segments, ...)
+
+    # clean-up of spatial coeffs
+    C = xp.nan_to_num(C, nan=0.0, posinf=0.0, neginf=0.0)
+
+    # back to torch if required
+    if is_torch:
+        B = _to_torch(B)
+        C = _to_torch(C)
+
+    return B, C
+
+
+def _outer_sum(xx, yy):
+    xx = xx[:, None, ...]  # add a singleton dimension at axis 1
+    yy = yy[None, ...]  # add a singleton dimension at axis 0
+    ss = xx + yy  # compute the outer sum
+    return ss
+
+
+# TODO: /* refactor with_* decorators
+def _to_numpy(input):
+    xp = get_array_module(input)
+
+    if xp.__name__ == "torch":
+        return input.numpy(force=True)
+    elif xp.__name__ == "cupy":
+        return input.get()
+    else:
+        return input
+
+
+def _to_cupy(input):
+    return cp.asarray(input)
+
+
+def _to_torch(input):
+    xp = get_array_module(input)
+
+    if xp.__name__ == "numpy":
+        return torch.from_numpy(input)
+    elif xp.__name__ == "cupy":
+        return torch.from_dlpack(input)
+    else:
+        return input
+
+
+# */
+
+
+class MRIFourierCorrected(FourierOperatorBase):
+    """Fourier Operator with B0 Inhomogeneities compensation.
+
+    This is a wrapper around the Fourier Operator to compensate for the
+    B0 inhomogeneities  in the  k-space.
+
+    Parameters
+    ----------
+    fourier_op: object of class FourierBase
+        the fourier operator to wrap
+    fieldmap : np.ndarray, optional
+        Rate map defined as ``fieldmap = R2*_map + 1j * B0_map``.
+        ``*_map`` and ``t`` should have reciprocal units.
+        If ``zmap`` is real, assume ``zmap = B0_map``.
+        Also supports Cupy arrays and Torch tensors.
+        Expected shape is ``(nz, ny, nx)``.
+    readout_time : np.ndarray or GPUarray, optional
+        Readout time in ``[s]`` of shape ``(nshots, npts)`` or ``(nshots * npts,)``.
```

Comment:
```suggestion
        Readout time in ``[s]`` of shape ``(n_shots, n_pts)`` or ``(n_shots * n_pts,)``.
```

Please keep a consistent naming convention  as in `n_time_segments` or `n_bins`

Reply:
Done, thanks!